compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -4586,6 +4586,58 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
             _ => "/* value */".to_string(),
         })
     }
+
+    fn suggest_add_result_as_return_type(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut Diag<'_>,
+        trait_ref: ty::PolyTraitRef<'tcx>,
+    ) {
+        if ObligationCauseCode::QuestionMark != *obligation.cause.code().peel_derives() {
+            return;
+        }
+
+        let node = self.tcx.hir_node_by_def_id(obligation.cause.body_id);
+        if let hir::Node::Item(item) = node
+            && let hir::ItemKind::Fn(sig, _, body_id) = item.kind
+            && let hir::FnRetTy::DefaultReturn(ret_span) = sig.decl.output
+            && trait_ref.skip_binder().args.len() == 2
+            && let ty::Tuple(l) = trait_ref.skip_binder().args.type_at(0).kind()
+            && l.len() == 0
+            && let ty::Adt(def, _) = trait_ref.skip_binder().args.type_at(1).kind()
+            && self.tcx.is_diagnostic_item(sym::Result, def.did())
+        {
+            let body = self.tcx.hir().body(body_id);
+            let mut sugg_spans =
+                vec![(ret_span, " -> Result<(), Box<dyn std::error::Error>>".to_string())];
+
+            if let hir::ExprKind::Block(b, _) = body.value.kind
+                && b.expr.is_none()
+                && let Some(s) = b.stmts.last()
+                && !is_ret_expr(s)
+            {
+                sugg_spans.push((
+                    s.span.shrink_to_hi().until(body.value.span.shrink_to_hi()),
```

Comment:
Yea I think this erases all the code after the `?` using statement (try adding a dummy `println!();` after the `?` using statement to test this.).

I would have expected for `b.span.shrink_to_high()` to suffice, but that may be after the `}`, so you may need to look for the last statement in the block (we know there must be one, as there's a `?`)

Reply:
if that span does not contain the `}`, please document it on the `span` field, so future readers immediately know what the span means.

Reply:
> Yea I think this erases all the code after the `?` using statement (try adding a dummy `println!();` after the `?` using statement to test this.).
> 
> I would have expected for `b.span.shrink_to_high()` to suffice, but that may be after the `}`, so you may need to look for the last statement in the block (we know there must be one, as there's a `?`)

Thank you very much. When I adding `println!();` as the last stmt of the block, I find the span of the stmt is `rust\library\std\src\macros.rs:83:23: 85:6`, so I think only use the span of the block is more appropriate.